homeassistant/components/bmp280/sensor.py

```diff
@@ -0,0 +1,161 @@
+        ),
+    }
+)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the sensor platform."""
+    try:
+        # this throws an exception if board is not supported
+        import board
+
+        # initializing I2C bus using the auto-detected pins
+        i2c = I2C(board.SCL, board.SDA)
+        # initializing the sensor
+        bmp280 = Adafruit_BMP280_I2C(i2c, address=config.get(CONF_I2C_ADDRESS))
+        # use custom name if there's any
+        name = config.get(CONF_NAME)
+        # BMP280 has both temperature and pressure sensing capability
+        add_entities(
+            [Bmp280TemperatureSensor(bmp280, name), Bmp280PressureSensor(bmp280, name)]
+        )
+    except NotImplementedError as error:
+        # this is thrown right after the import statement if the board is unsupported
+        if error.args[0] == "Board not supported":
+            _LOGGER.error(
+                "Failed to determine board type. Is this instance running on a Raspberry Pi or a similar I2C capable device?"
+            )
+            raise PlatformNotReady()
+        raise error
+    except ValueError as error:
+        # this happens when the board is I2C capable, but the device is not found at the configured address
+        if str(error.args[0]).startswith("No I2C device at address"):
+            _LOGGER.error(error.args[0])
+            raise PlatformNotReady()
+        raise error
+
+
+class Bmp280Sensor(Entity):
+    """Base class for BMP280 entities."""
+
+    errored = False
+
+    def __init__(
+        self,
+        bmp280: Adafruit_BMP280_I2C,
+        name: str,
+        unit_of_measurement: str,
+        device_class: str,
+    ):
+        """Initialize the sensor."""
+        self._bmp280 = bmp280
+        self._name = name
+        self._unit_of_measurement = unit_of_measurement
+        self._device_class = device_class
+        self._state = None
+        self._errored = False
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        return self._unit_of_measurement
+
+    @property
+    def device_class(self):
+        """Return the device class."""
+        return self._device_class
+
+    @property
+    def available(self) -> bool:
+        """Return if the device is currently available."""
+        return not Bmp280Sensor.errored
+
+
+class Bmp280TemperatureSensor(Bmp280Sensor):
+    """Representation of a Bosch BMP280 Temperature Sensor."""
+
+    def __init__(self, bmp280: Adafruit_BMP280_I2C, name: str):
+        """Initialize the entity."""
+        super().__init__(
+            bmp280, f"{name} Temperature", TEMP_CELSIUS, DEVICE_CLASS_TEMPERATURE
+        )
+
+    def update(self):
```

Comment:
Do updates need a Throttle?

Reply:
I'm afraid I don't understand what do you mean by this. I have searched the development documentation for the keyword *throttle* but came up empty.

Reply:
The BME280 integration has an example.
```
class BME280Handler:
    """BME280 sensor working in i2C bus."""

    def __init__(self, sensor):
        """Initialize the sensor handler."""
        self.sensor = sensor
        self.update(True)

    @Throttle(MIN_TIME_BETWEEN_UPDATES)
    def update(self, first_reading=False):
        """Read sensor data."""
        self.sensor.update(first_reading)
```

Reply:
Excellent answer, I'll check if it's needed. 

If I'd had to guess I'd say querying from an i2c sensor doesn't use a lot of resources and the chip is probably also capable of answering queries pretty fast, but I'll check the data sheet for a definite answer

Reply:
According to the [datasheet](https://www.bosch-sensortec.com/media/boschsensortec/downloads/datasheets/bst-bmp280-ds001.pdf) the chip is capable of sampling data at a rate of up to 157 Hz. The typical value of the sampling rate in the slowest (most accurate) mode is 26.7 Hz (chapter 3.8.1). It has shadowing, so it's possible to read the previous value while it's still measuring the next one-

Also based on the information in chapter 5.4.2 an I2C clock cycle can be as fast as 200ns, so I'm quite confident that home assistant is not likely to overload this sensor with its requests.

However with all this being said I can still add a `Throttle` decorator just to be safe if it doesn't imply a noticable performance penalty. What do you think?

Reply:
Nevermind, I'll just use the same settings as BME280 so it's more consistent anyways

Reply:
Sounds good. Is there ever a possibility that two home assistants could be polling the same sensor?

Reply:
That's a good point, I've never considered the fact that running two instances in parallel is even possible.

However since I2C is a serial protocol I'd imagine the operating system somehow implements exclusive access to the port so separate instances should be able to access the same device without major problems. But to make this sure one has to dive down through the `Adafruit_BMP280_I2C` - `busio` - who-knows-what-else - `wire` chain which looks like a bit of an overkill :)